docs: Add docs for GenSX Cloud (#589)
gensx-inc/gensx: website/docs/src/content/cloud/observability.mdx

---
title: Observability & tracing
description: Monitor and debug your GenSX workflows
---

# Observability & tracing

> GenSX provides observability tools that make it easy to understand, debug, and optimize your workflows. Every component execution is automatically traced, giving you full visibility into what's happening inside your LLM workflows. You can view traces in realtime as workflows execute, and view historical traces to debug production issues like hallucinations.

Reviewer: Should we add a "when logged in" caveat?

Author: eh, this section is under cloud so I this should be implicitly understood.

## Viewing traces

When you run a workflow, GenSX automatically generates a trace that captures the entire execution flow, including all component inputs, outputs, and timing information.

### Accessing the trace viewer

The GenSX cloud console includes a trace viewer. You can access traces in several ways:

1. **From the Console**: Navigate to your project in the [GenSX Console](https://app.gensx.com) and select the "Executions" tab
2. **Trace URL**: When running a workflow with `printUrl: true`, a direct link to the trace is printed to the console
3. **API Response**: The execution ID from API responses can be used to view traces in the console

```tsx
// Executing a workflow with trace URL printing
const result = await MyWorkflow.run(
  { input: "What is GenSX?" },
  { printUrl: true },
);

// Console output includes:

> // [GenSX] View execution at: https://app.gensx.com/your_org/executions/your_execution_idf

Reviewer: ```suggestion
// [GenSX] View execution at: https://app.gensx.com/your_org/executions/your_execution_id
```

```

> ### Viewing the flame graph

Reviewer: "Viewing" feels repetitive since you just told people how to access it. the section is more explaining how to read the graph too

```suggestion
### Understanding the flame graph
```

The flame graph visualizes the entire execution tree including branches, all nested sub-components, and timing:

![Workflow component tree](/cloud/trace.png)

- **Component hierarchy**: See the nested structure of your components and their relationships
- **Execution timing**: The width of each bar represents the relative execution time
- **Status indicators**: Quickly spot errors or warnings with color coding
- **Component filtering**: Focus on specific components or component types

Click on any component in the flame graph to inspect its details, including inputs, outputs, and timing information.

### Viewing component inputs and outputs

For each component in your workflow, you can inspect:

1. **Input properties**: All props passed to the component
2. **Output values**: The data returned by the component
3. **Execution duration**: How long the component took to execute

> 4. **Metadata**: Additional information like token counts for LLM calls

Reviewer: Is this technically true? token counts are only there if they're in the component output (for example <ChatCompletion> won't have token details)

Author: Will leave this as I expect we'll improve it in the near future.

![Component trace](/cloud/component-trace.png)

This visualization is particularly valuable for debugging production and user-reported issues like hallucinations.

### Viewing historical traces

The GenSX Console maintains a history of all your workflow executions, allowing you to:

- **Compare executions**: See how behavior changes across different runs
- **Identify patterns**: Spot recurring issues or performance bottlenecks
- **Filter by status**: Focus on successful, failed, or in-progress executions
- **Search**: Find historical executions

Historical traces are automatically organized by project and environment, making it easy to find relevant executions.

## Configuring traces

> tGenSX provides flexible options for configuring and organizing traces for the GenSX Cloud serverless platform, local development, and any other deployment platform like vercel, cloudflare and AWS.

Reviewer: ```suggestion
GenSX provides flexible options for configuring and organizing traces for the GenSX Cloud serverless platform, local development, and any other deployment platform like vercel, cloudflare and AWS.
```

### Tracing GenSX Cloud workflows

> When running workflows on GenSX Cloud, tracing is automatically configured:

Reviewer: Sentence is a bit ambiguous as is

```suggestion
When running workflows deployed to the GenSX Cloud, tracing is automatically configured:
```

- **Project context**: Traces are associated with the correct project
- **Environment segregation**: Development, staging, and production traces are kept separate
- **Authentication**: API keys and organization information are handled automatically
- **Retention**: Traces are stored according to your plan limits

No additional configuration is needed – everything works out of the box.

### Tracing on other deployment platforms

> To enable tracing for workflows deployed outside of GenSX Cloud (like AWS Lambda, Google Cloud Functions, etc.), you need to set several environment variables:

Reviewer: Fixed the name of GCP Cloud Run (if it's too long "Azure Functions" would be shorter :))

```suggestion
To enable tracing for workflows deployed outside of GenSX Cloud (like AWS Lambda, GCP Cloud Run functions, etc.), you need to set several environment variables:
```

```bash
# Required variables
GENSX_API_KEY=your_api_key_here
GENSX_ORG=your_gensx_org_name
GENSX_PROJECT=your_project_name

# Optional variables
GENSX_ENVIRONMENT=your_environment_name        # Separate traces into specific environments
GENSX_CHECKPOINTS=false                        # Explicitly disable
```

### Configuring traces for local development

> For local development, tracing configuration is automatically inferred from:

Reviewer: ```suggestion
For local development, the tracing configuration is automatically inferred from:
```

1. The `gensx.yaml` file in your project root
2. Your local configuration managed by the `gensx` CLI in `~/.config/gensx/config`

> 3. Optionally the `GENSX_ENVIRONMENT` environment variable to separate local traces from other environments

Reviewer: ```suggestion
3. Optionally the `GENSX_ENVIRONMENT` environment variable can be set to separate local traces from other environments
```

Does this actually work though? do we create the environment for you when you do this?

Author: Outstanding TODO we need to fix before launch: https://github.com/gensx-inc/gensx/issues/583

The local development server started with `gensx start` uses this same configuration scheme as well.

### Organizing traces by environment

> GenSX allows you to organize traces by environment (development, staging, production) to keep your debugging data well-structured:

Reviewer: It's not super clear reading this that you have control over your environments and that we don't automatically give you dev, staging, prod. Maybe we need an extra sentence stating that?

### Marking component outputs as secrets

For components that might return sensitive information, you can mark the entire output as sensitive:

```tsx
const GenerateCredentials = gensx.Component(
  "GenerateCredentials",
  async ({ userId }) => {
    // This entire output will be marked as secret
    return {
      accessToken: "sk-1234567890abcdef",
      refreshToken: "rt-0987654321fedcba",
      expiresAt: Date.now() + 3600000,
    };
  },
  {
    secretOutputs: true,
  },
);
```

When `secretOutputs` is set to `true`, the entire output object or value will be treated as sensitive and masked in traces.

### Secrets propagation

Secrets are automatically propagated through the component hierarchy:

```tsx
const ParentComponent = gensx.Component(
  "ParentComponent",
  async ({ apiKey, query }) => {
    // apiKey is marked as secret
    const results = await ChildComponent.run({
      query,
      headers: { authorization: `Bearer ${apiKey}` },
    });

    // apiKey is also scrubbed here in the trace,
    // even though ChildComponent doesn't explicitly mark it
    return processResults(results, apiKey);
  },
  {
    secretProps: ["apiKey"],
  },
);
```

> Secrets remain scrubbed throughout the execution chain, even if they're transformed or embedded in different objects.

Reviewer: is this section necessary? I'm not sure if it adds any clarity beyond this sentence from above:  "Even if a secret is passed down through multiple components, it remains scrubbed in all traces."

I personally found this sentence very confusing: Secrets are automatically propagated through the component hierarchy:

Author: removed.
